### PR TITLE
Fix test generation

### PR DESCRIPTION
**Pull request by QuentinBisson**
So reviewing this will be a hassle :(

This PR fixes tests generated files to support both capi and vintage providers separately.

The main reason behind this change is that current code actually rendered CAPI clusters on vintage MCs and vintage WCs for CAPI MC which should never happen.

## Checklist

I have:

- [x] Described why this change is being introduced
- [x] Separated out refactoring/reformatting in a dedicated PR
- [x] Updated changelog in `CHANGELOG.md`

### CODE REVIEW

File: pkg/unittest/input/capi/case-1-capa-mc.golden [hunk 1]
```diff
@@ -0,0 +1,11 @@
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: Cluster
+metadata:
+  name: test-installation
```
Comment (QuentinBisson) [inline pkg/unittest/input/capi/case-1-capa-mc.golden:4]: This template is a CAPI MC

File: pkg/unittest/input/capi/case-1-capa-mc.golden [hunk 2]
```diff
@@ -0,0 +1,11 @@
+apiVersion: cluster.x-k8s.io/v1beta1
```
Comment (QuentinBisson) [inline pkg/unittest/input/capi/case-1-capa-mc.golden:1]: All inputs under capi are CAPI clusters only

File: pkg/unittest/input/capi/case-3-capz.golden [hunk 3]
```diff
@@ -0,0 +1,11 @@
+apiVersion: cluster.x-k8s.io/v1beta1
```
Comment (QuentinBisson) [inline pkg/unittest/input/capi/case-3-capz.golden:1]: CAPZ WC

File: pkg/unittest/input/capi/case-2-capa.golden [hunk 4]
```diff
@@ -0,0 +1,11 @@
+apiVersion: cluster.x-k8s.io/v1beta1
```
Comment (QuentinBisson) [inline pkg/unittest/input/capi/case-2-capa.golden:1]: CAPA WC

File: pkg/unittest/input/capi/case-4-eks.golden [hunk 5]
```diff
@@ -1,8 +1,6 @@
 apiVersion: cluster.x-k8s.io/v1beta1
```
Comment (QuentinBisson) [inline pkg/unittest/input/capi/case-4-eks.golden:1]: EKS WC

File: pkg/unittest/input/capi/case-5-gcp.golden [hunk 6]
```diff
@@ -0,0 +1,11 @@
+apiVersion: cluster.x-k8s.io/v1beta1
```
Comment (QuentinBisson) [inline pkg/unittest/input/capi/case-5-gcp.golden:1]: GCP WC

File: pkg/unittest/unittest.go [hunk 7]
```diff
@@ -129,7 +143,11 @@ func (r *Runner) Run() error {
 
 			outputFilePath := filepath.Join(r.OutputDir, file.Name())
 			if r.Update {
-				err := os.WriteFile(outputFilePath, testResult, 0644) // #nosec
+				err := os.MkdirAll(r.OutputDir, os.ModePerm)
```
Comment (QuentinBisson) [inline pkg/unittest/unittest.go:146]: Creates the whole golden files path to smooth the tests up

File: service/controller/clusterapi/controller.go [hunk 8]
```diff
@@ -61,9 +60,9 @@ func NewController(config ControllerConfig) (*Controller, error) {
 
 	var resources []resource.Interface
 	{
-		c := controllerresource.Config(config)
+		c := Config(config)
```
Comment (QuentinBisson) [inline service/controller/clusterapi/controller.go:63]: Just moved the resource file

File: service/controller/resource/alerting/alertmanagerconfig/resource_test.go [hunk 9]
```diff
@@ -16,33 +16,37 @@ func TestRenderingOfAlertmanagerNotificationTemplate(t *testing.T) {
...
 		testFunc = func(v interface{}) (interface{}, error) {
 			return renderNotificationTemplate(unittest.ProjectRoot(), config)
 		}
 	}
 
-	outputDir, err := filepath.Abs("./test/notification-template")
-	if err != nil {
-		t.Fatal(err)
-	}
+	for _, flavor := range unittest.ProviderFlavors {
```
Comment (QuentinBisson) [inline service/controller/resource/alerting/alertmanagerconfig/resource_test.go:27]: All tests are now run for the 2 providers separately

File: service/controller/resource/alerting/alertmanagerconfig/resource_test.go [hunk 10]
```diff
@@ -16,33 +16,37 @@ func TestRenderingOfAlertmanagerNotificationTemplate(t *testing.T) {
...
-	}
-	runner, err := unittest.NewRunner(c)
-	if err != nil {
-		t.Fatal(err)
-	}
+		c := unittest.Config{
+			OutputDir:            outputDir,
+			T:                    t,
+			TestFunc:             testFunc,
+			Flavor:               flavor,
```
Comment (QuentinBisson) [inline service/controller/resource/alerting/alertmanagerconfig/resource_test.go:37]: This is where we pass the flavor

File: service/controller/resource/monitoring/prometheus/test/capi/case-4-eks.golden [hunk 11]
```diff
@@ -37,7 +37,7 @@ spec:
     customer: Giant Swarm
     installation: test-installation
     pipeline: testing
-    provider: aws
+    provider: eks
```
Comment (QuentinBisson) [inline service/controller/resource/monitoring/prometheus/test/capi/case-4-eks.golden:40]: EKS as expected

File: service/controller/resource/monitoring/prometheus/resource_test.go [hunk 12]
```diff
@@ -36,69 +32,78 @@ func TestPrometheus(t *testing.T) {
...
+
+				config := Config{
+					Address:            "http://prometheus/cluster",
+					Customer:           "Giant Swarm",
+					EvaluationInterval: "60s",
+					Installation:       "test-installation",
+					Pipeline:           "testing",
+					K8sClient:          k8sClient,
+					Provider: cluster.Provider{
+						Kind:   "aws",
```
Comment (QuentinBisson) [inline service/controller/resource/monitoring/prometheus/resource_test.go:84]: Provider is now correct for CAPI and vintage stays the same

File: service/controller/resource/monitoring/scrapeconfigs/resource_test.go [hunk 13]
```diff
@@ -135,11 +135,12 @@ func TestAWSScrapeconfigs(t *testing.T) {
 	}
 
 	c := unittest.Config{
+		Flavor:               "vintage",
```
Comment (QuentinBisson) [inline service/controller/resource/monitoring/scrapeconfigs/resource_test.go:138]: Setting vintage flavor for aws and azure tests but set capi flavor for CAPI MCs

File: pkg/unittest/input/capi/case-1-capa-mc.golden (hunk 1 above)
Comment (hervenicol) [reply pkg/unittest/input/capi/case-1-capa-mc.golden:4]: I think if your code deserve some comments, they should rather be in the code than in the PR :stuck_out_tongue_closed_eyes:

Comment (QuentinBisson) [reply pkg/unittest/input/capi/case-1-capa-mc.golden:4]: Well I was trying to explain that to you but yes you're right, I'll make this more readable first

File: service/controller/resource/monitoring/prometheus/resource_test.go (hunk 12 above)
Comment (QuentinBisson) [reply service/controller/resource/monitoring/prometheus/resource_test.go:84]: So it should not but with our change,we do not even use the kind anymore, this a remnant of vintage

Comment (QuentinBisson) [reply service/controller/resource/monitoring/prometheus/resource_test.go:84]: I could make this capa if flavor is capi, I just did not feel the need

File: service/controller/resource/monitoring/verticalpodautoscaler/resource_test.go [hunk 14]
```diff
@@ -18,14 +18,10 @@ import (
 	"github.com/giantswarm/prometheus-meta-operator/v2/pkg/unittest"
 )
 
-var update = flag.Bool("update", false, "update the output file")
+var update = flag.Bool("update", false, "update the ouput file")
```
Comment (hervenicol) [inline service/controller/resource/monitoring/verticalpodautoscaler/resource_test.go:21]: I guess the `t` was deleted by accident?
```suggestion
var update = flag.Bool("update", false, "update the output file")
```

Comment (QuentinBisson) [reply service/controller/resource/monitoring/verticalpodautoscaler/resource_test.go]: Well, yes I fixed them all :)

**Review by hervenicol — APPROVED** [review]
My eyes are sore after reading all that :rofl: 


But I think that's a great improvement, especially since it's been a long time we have lots of CAPI-specific code in PMO and tests did not show it.

We may still find some places where the tests require improvement in the future, but this PR is already great (in all meanings :rofl: )

**Comment by QuentinBisson** [comment]
I will only say thank you for reviewing this monster

**Pull request merged**